### PR TITLE
[Feature] Impl insert ignore stmt for primary key

### LINKED ISSUE #46259: Support INSERT IGNORE Statement for Primary Key Table
## Feature request
Support INSERT IGNORE Statement for Primary Key Table

Insert ignore is a very common use in mysql community.
https://dev.mysql.com/doc/refman/8.0/en/insert.html

grammar:
```SQL
INSERT IGNORE
    INTO table_name 
    [( column_name [, …] )]
    [VALUES]
    [(value_list[, …])]
    [query];
```

If where are same primary keys, default insert will use the newer row replace the old row but the insert ignore stmt will ignore the newer row and keep the old row.

So `INSERT IGNORE` was only enabled in Primary Key table.

Also the other kinds of data load such as StreamLoad and Broker Load should be suppported.

### PR DESCRIPTION
**Pull request by SaintBacchus**
## Why I'm doing:
Fixes #46259 

## What I'm doing:
Support insert ignore for dml stmt and StreamLoad

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by imay** [comment]
@SaintBacchus can you create a GitHub issue to describe this feature clearly?

**Comment by SaintBacchus** [comment]
OK, I will write a issue later

> @SaintBacchus can you create a GitHub issue to describe this feature clearly?

File: be/src/storage/lake/update_manager.cpp [hunk 1]
```diff
@@ -206,6 +206,10 @@ Status UpdateManager::publish_primary_key_tablet(const TxnLogPB_OpWrite& op_writ
     for (uint32_t segment_id = 0; segment_id < op_write.rowset().segments_size(); segment_id++) {
         new_deletes[rowset_id + segment_id] = {};
     }
+    InsertMode insert_mode = InsertMode::UPSERT;
+    if (op_write.insert_mode()) {
+        insert_mode = InsertMode::IGNORE;
```
Comment (dujl) [inline be/src/storage/lake/update_manager.cpp:211]: 这里应该是insert_mode = op_write.insert_mode(); ？

File: be/src/storage/tablet_updates.cpp [hunk 2]
```diff
@@ -1262,6 +1262,10 @@ void TabletUpdates::_apply_normal_rowset_commit(const EditVersionInfo& version_i
     }
     EditVersion latest_applied_version;
     st = get_latest_applied_version(&latest_applied_version);
+    InsertMode insert_mode = InsertMode::UPSERT;
+    if (rowset->rowset_meta()->get_meta_pb_without_schema().insert_mode()) {
```
[the remark this answers is not in the document]
Comment (SaintBacchus) [reply be/src/storage/tablet_updates.cpp:1266]: 这里漏改了

File: be/src/http/action/stream_load.cpp [hunk 3]
```diff
@@ -600,7 +600,15 @@ Status StreamLoadAction::_process_put(HttpRequest* http_req, StreamLoadContext*
     } else if (!http_req->header(HTTP_COMPRESSION).empty()) {
         request.__set_payload_compression_type(http_req->header(HTTP_COMPRESSION));
     }
-
+    if (!http_req->header(HTTP_INSERT_MODE).empty()) {
```
Comment (dujl) [inline be/src/http/action/stream_load.cpp:603]: if there is no HTTP_INSERT_MODE header,  invoke empty will rasie npe or not ?

File: be/src/http/action/transaction_stream_load.cpp [hunk 4]
```diff
@@ -436,7 +436,13 @@ Status TransactionStreamLoadAction::_parse_request(HttpRequest* http_req, Stream
     } else if (!http_req->header(HTTP_COMPRESSION).empty()) {
         request.__set_payload_compression_type(http_req->header(HTTP_COMPRESSION));
     }
-
+    if (!http_req->header(HTTP_INSERT_MODE).empty()) {
```
Comment (dujl) [inline be/src/http/action/transaction_stream_load.cpp:439]: Will calling empty cause an NPE wen there is no HTTP_INSERT_MODE?

File: be/src/http/action/stream_load.cpp (hunk 3 above)
Comment (SaintBacchus) [reply be/src/http/action/stream_load.cpp:603]: Same with other header

**Comment by jaogoy** [comment]
@decster  Could you pay attention on this PR? And, is it compatible with you partial update function in the INSERT syntax?

File: gensrc/proto/olap_file.proto [hunk 5]
```diff
@@ -175,6 +183,8 @@ message RowsetMetaPB {
     optional uint32 max_compact_input_rowset_id = 61;
     // global transaction id
     optional int64 gtid = 62;
+    // insert mode
+    optional InsertMode insert_mode = 63 [default = UPSERT_MODE];
```
Comment (decster) [inline gensrc/proto/olap_file.proto:187]: why not put this in 
    // some txn semantic information bind to this rowset
    optional RowsetTxnMetaPB txn_meta = 55;

File: fe/fe-core/src/main/java/com/starrocks/planner/StreamLoadPlanner.java [hunk 6]
```diff
@@ -154,6 +155,17 @@ public TExecPlanFragmentParams plan(TUniqueId loadId) throws UserException {
...
+        if (streamLoadInfo.getInsertMode() == TInsertMode.IGNORE_MODE) {
+            if (!isPrimaryKey) {
+                throw new DdlException("Only primary key table support ignore insert mode");
+            }
+            if (streamLoadInfo.isPartialUpdate()) {
+                throw new DdlException("Ignore insert mode is not support partial update.");
+            }
+            if (streamLoadInfo.getMergeConditionStr() != null && !streamLoadInfo.getMergeConditionStr().isEmpty()) {
+                throw new DdlException("Ignore insert mode is not support condition update.");
+            }
```
Comment (decster) [inline fe/fe-core/src/main/java/com/starrocks/planner/StreamLoadPlanner.java:167]: ignore insert mode -> insert ignore mode ?
is not support -> does not support
`is not support condition update` -> `does not support conditional update`

File: gensrc/proto/olap_file.proto (hunk 5 above)
Comment (SaintBacchus) [reply gensrc/proto/olap_file.proto:187]: `txn_meta` was only created in update case
image

File: fe/fe-core/src/main/java/com/starrocks/planner/StreamLoadPlanner.java (hunk 6 above)
Comment (SaintBacchus) [reply fe/fe-core/src/main/java/com/starrocks/planner/StreamLoadPlanner.java]: fine

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/InsertAnalyzer.java [hunk 7]
```diff
@@ -342,7 +343,19 @@ private static Table getTargetTable(InsertStmt insertStmt, ConnectContext sessio
                 throw unsupportedException(msg);
             }
         }
-
+        if (insertStmt.isIgnore()) {
+            if (!(table instanceof OlapTable)) {
+                throw unsupportedException("Only support insert ignore for olap table");
```
Comment (jaogoy) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/InsertAnalyzer.java:348]: Put these error messages into ErrorCode.java.
And print some necessary fields for usability, here we can print table name and table type:
`Only support INSERT IGNORE for olap table, table=$name, type=$table_type`

File: be/src/http/action/stream_load.cpp [hunk 8]
```diff
@@ -600,7 +600,15 @@ Status StreamLoadAction::_process_put(HttpRequest* http_req, StreamLoadContext*
...
+        if (boost::iequals(http_req->header(HTTP_INSERT_MODE), "upsert")) {
+            request.__set_insert_mode(TInsertMode::type::UPSERT_MODE);
+        } else if (boost::iequals(http_req->header(HTTP_INSERT_MODE), "ignore")) {
+            request.__set_insert_mode(TInsertMode::type::IGNORE_MODE);
+        } else {
+            return Status::InvalidArgument("Invalid insert mode flag format. Must be upsert or ignore");
```
Comment (jaogoy) [inline be/src/http/action/stream_load.cpp:609]: Print the parameter value

File: be/src/http/action/transaction_stream_load.cpp [hunk 9]
```diff
@@ -436,7 +436,15 @@ Status TransactionStreamLoadAction::_parse_request(HttpRequest* http_req, Stream
...
```
Comment (jaogoy) [inline be/src/http/action/transaction_stream_load.cpp:439]: I wonder why don't we reuse the same code snippet?

File: be/src/storage/aggregate_type.h [hunk 10]
```diff
@@ -29,7 +29,9 @@ enum StorageAggregateType {
     STORAGE_AGGREGATE_BITMAP_UNION = 7,
     // Replace if and only if added value is not null
     STORAGE_AGGREGATE_REPLACE_IF_NOT_NULL = 8,
-    STORAGE_AGGREGATE_PERCENTILE_UNION = 9
+    STORAGE_AGGREGATE_PERCENTILE_UNION = 9,
+    // used in insert ignore mode to keep fist value.
+    STORAGE_AGGREGATE_FIRST = 10
```
Comment (jaogoy) [inline be/src/storage/aggregate_type.h:34]: What does `first` mean? The last version value? Or the first random row with the same pk in a INSERT batch?

**Review by jaogoy — COMMENTED** [review]
I'm not fimilar with the code,but I'm really confused that so many files are modified.
And, I wonder is there some ways to optimize it? such as:
1. A parameter can be passed through a whole object containing the new parameter rather than adding a single parameter in a lot of places.
2. Only the final exec code will take response on this parameter.

File: be/src/storage/aggregate_type.h (hunk 10 above)
Comment (SaintBacchus) [reply be/src/storage/aggregate_type.h:34]: like agg function `first_value`

**Comment by SaintBacchus** [comment]
> I'm not fimilar with the code,but I'm really confused that so many files are modified. And, I wonder is there some ways to optimize it? such as:
> 
> 1. A parameter can be passed through a whole object containing the new parameter rather than adding a single parameter in a lot of places.
> 2. Only the final exec code will take response on this parameter.

This feature adds a new stmt and a new property for stream load and this property will go through the execution layer and storage layer, and finally add property to the primary key index. 
There is no interface to add this property directly to the primary key.

File: be/src/http/action/transaction_stream_load.cpp (hunk 9 above)
Comment (SaintBacchus) [reply be/src/http/action/transaction_stream_load.cpp:439]: just keep the same code style near this code.

**Pull request closed without merging**